### PR TITLE
MO-957: New responsiblity rules for ISP & Parole

### PR DESCRIPTION
**Pull request by ryantk**
Ticket: https://dsdmoj.atlassian.net/jira/software/c/projects/MO/boards/518?selectedIssue=MO-957

### CODE REVIEW

File: app/models/hmpps_api/offender_sentence_term/sentence_type.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      # 14FTR_ORA     ORA 14 Day Fixed Term Recall
+      # LR_ORA        ORA Licence Recall
+      # ADIMP_ORA     ORA Sentencing Code Standard Determinate Sentence
+      # LR_ALP        Recall from Automatic Life
+      # LR_ALP_CDE21  Recall from Automatic Life Sec 283 Sentencing Code (21+)
+      # LR_DLP        Recall from Discretionary Life
+      # LR_LIFE       Recall to Custody Indeterminate Sentence
+      # ADIMP         Sentencing Code Standard Determinate Sentence
+
+      INDETERMINATE_SENTENCE_TYPES = [
```
Comment (ryantk) [inline app/models/hmpps_api/offender_sentence_term/sentence_type.rb:28]: When looking up sentence terms, if the sentence type is in this list it is considered an indeterminate sentence

File: app/models/calculated_handover_date.rb [hunk 2]
```diff
@@ -59,6 +59,11 @@ def community_supporting?
     responsibility.in? [CUSTODY_WITH_COM]
   end
 
+  alias_method :pom_responsible?, :custody_responsible?
+  alias_method :pom_supporting?, :custody_supporting?
+  alias_method :com_responsible?, :community_responsible?
+  alias_method :com_supporting?, :community_supporting?
```
Comment (ryantk) [inline app/models/calculated_handover_date.rb:65]: Making the methods match the language used in the requirements

File: app/lib/handover/handover_calculation.rb [hunk 3]
```diff
@@ -14,10 +14,12 @@ def calculate_handover_date(sentence_start_date:,
...
+      elsif sentence_start_date + 10.months >= earliest_release_date
+        [nil, :determinate_short]
+      elsif is_determinate_parole
+        [earliest_release_date - 12.months, :determinate_parole]
       else
-        calculate_determinate_handover_date(sentence_start_date: sentence_start_date,
-                                            earliest_release_date: earliest_release_date,
-                                            is_determinate_parole: is_determinate_parole)
+        [earliest_release_date - 8.months - 14.days, :determinate]
       end
```
Comment (ryantk) [inline app/lib/handover/handover_calculation.rb:23]: Inlined the `calculate_determinate_handover_date` method as it was harder to understand that way. And it was not used elsewhere.

File: app/models/mpc_offender.rb [hunk 4]
```diff
@@ -174,6 +174,10 @@ def target_hearing_date
     end
   end
 
+  def ppud_or_manually_entered_target_hearing_date
+    most_recent_parole_review&.target_hearing_date || @offender.parole_record&.target_hearing_date
+  end
```
Comment (ryantk) [inline app/models/mpc_offender.rb:179]: Manually entered THD OR THD from PPUD

File: app/models/mpc_offender.rb [hunk 5]
```diff
@@ -237,6 +241,12 @@ def due_for_release?
     most_recent_parole_review&.current_record_hearing_outcome == 'Release'
   end
 
+  def sentenced_to_an_additional_isp?
+    @sentenced_to_an_additional_isp ||= OffenderService.get_offender_sentences_and_offences(booking_id)
+      .select(&:indeterminate?)
+      .count > 1
+  end
```
Comment (ryantk) [inline app/models/mpc_offender.rb:248]: If more than 1 indeterminate sentence is within the current booking

File: app/models/mpc_offender.rb [hunk 6]
```diff
@@ -477,6 +469,10 @@ def handover
                   end
   end
 
+  def responsibility
+    @offender.responsibility || @offender.calculated_handover_date || HandoverDateService.handover(self)
+  end
```
Comment (ryantk) [inline app/models/mpc_offender.rb:474]: Manually entered responsibility or pre-calculated responsibility or generate new handover now

File: app/services/handover_date_service.rb [hunk 7]
```diff
@@ -29,33 +29,50 @@ def self.handover(mpc_offender)
...
         earliest_release_date: offender.earliest_release.date,
         is_early_allocation: offender.early_allocation?,
         is_indeterminate: offender.indeterminate_sentence?,
         in_open_conditions: offender.in_open_conditions?,
         is_determinate_parole: offender.determinate_parole?,
       )
-      handover_start_date = Handover::HandoverCalculation.calculate_handover_start_date(
-        handover_date: handover_date,
+      start_date = Handover::HandoverCalculation.calculate_handover_start_date(
+        handover_date:,
```
Comment (doctorpod) [inline app/services/handover_date_service.rb:75]: Is this meant to be nil?

**Review by doctorpod — APPROVED** [review]
Approved but 1 query

File: app/services/handover_date_service.rb (hunk 7 above)
Comment (ryantk) [reply app/services/handover_date_service.rb:75]: @doctorpod it's hash value omission syntax, came out in ruby 3.1
https://bugs.ruby-lang.org/issues/14579

**Pull request merged**